Lib/multiprocessing/shared_memory.py

```diff
@@ -239,6 +240,9 @@ def unlink(self):
         In order to ensure proper cleanup of resources, unlink should be
         called once (and only once) across all processes which have access
         to the shared memory block."""
+        if not self._track:
+            raise TypeError("unlink() must be called by a tracking instance")
```

Comment:
This seems not promising due to 2 facts:

1. It's not a `TypeError` because tracking or not is not reflected on the type itself.
2. This makes `unlink`-ing a shared memory impossible. I suggest doing the `shm_unlink` but not calling `unregister` if not tracked.

Reply:
+1 with both remarks.

Reply:
Can you elaborate? What would the right error be? Unlinking the memory is perfectly possible by using `track=True` which is even the default. 

Let's follow your suggestion, then a typical scenario may unfold like this:

1. Some main process creates shared memory and hands over the name to some subprocess.
2. The subprocess attaches without tracker but unlinks the memory. 
3. The main process terminates and its resource tracker will first complain about the lingering shared memory and then print an exception message about being unable to delete a file.

If we are fine with the warnings by the resource tracker then you're right that the error is not needed. If we see the warnings as something that the user should fix in the code, then the error helps the user discover the issue possibly much earlier. Though I do see the point that an error is too extreme for something that would result in a warning eventually.

Reply:
I will undo this part and add a check to the resource tracker to not complain if the user did clean up, after all.

Reply:
The new `unlink` looks good to me!